TELCODOCS#2637: Adding custom MCP upgrade content (#111682)
openshift/openshift-docs: modules/configuring-custom-machine-config-pools-parallel-upgrades.adoc

.Procedure

. Create a YAML file for each custom MCP that corresponds to a failure domain in your cluster, as in the following example:
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-0
  labels:
    machineconfiguration.openshift.io/role: worker-0
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values: [ worker, worker-0 ]
  paused: true
  maxUnavailable: 100%
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/worker-0: ""
----
+
Ensure the following configurations are present:

* *maxUnavailable*: Set this value to `100%`. This setting enables the Machine Config Operator (MCO) to update all nodes in the single failure domain simultaneously, significantly reducing upgrade time for the zone.

* *paused*: Set this value to `true`. Pausing the pool prevents unintended updates from starting immediately. You unpause the pool only when you are ready to upgrade that specific zone.

* *nodeSelector*: Define a label to identify the nodes that belong to this pool (for example, `node-role.kubernetes.io/worker-0`).

> . Apply the `topology.kubernetes.io/zone` label to identify the KFD for the Kubernetes scheduler, and the custom node role label (for example, `worker-0`) to assign the node to the MCP by running the following command:

Reviewer: Apply the.... to identify **each** KFD for the Kubernetes scheduler.

Looks like only one topology label needs to be added, but a different one must be added to each MCP (worker-0, worker-1, worker-2, worker-3..).

Author: Updated as follows:

. Apply the `topology.kubernetes.io/zone` label to each node to identify its KFD for the scheduler. You must apply the corresponding custom node role label (for example, `worker-0`, `worker-1`, `worker-2`) to assign each node to its custom MCP by running the following command:

Reviewer: Forgot to mention that KFD stands for Kubernetes Failure Domain. I think it would be good to include it since KFD could not be easy to find out.

I would say: Apply the topology.kubernetes.io/zone label to each node to identify its Kubernetes failure domain (KFD) for the scheduler.

Author: Updated as per the suggestion.

> . Apply the `topology.kubernetes.io/zone` label to identify the KFD for the Kubernetes scheduler, and the custom node role label (for example, `worker-0`) to assign the node to the MCP by running the following command:

Reviewer: I would recommend including the node role label on each node during installation so you avoid any disruption on the node once the cluster is installed.  The label, then, will already be included when the cluster is ready.

It can also be done later in a cluster already installed, but a warning message should be included that moving a node from a different MCP can cause disruption.

Author: Updated the IMPORTANT NOTE to include this new info:

"Apply the `topology.kubernetes.io/zone` label and custom node role labels during cluster installation or node scaling whenever possible. Applying these labels before scheduling workloads ensures that the Kubernetes scheduler can distribute application replicas correctly across failure domains.

Applying or changing node role labels after installation can move nodes between MCPs, which might temporarily disrupt workloads on those nodes. If workloads are already running when you apply or modify these labels, you might need to reschedule workloads to achieve the intended HA distribution."

:_mod-docs-content-type: PROCEDURE
[id="configuring-custom-machine-config-pools-parallel-upgrades_{context}"]
= Configuring custom machine config pools for parallel upgrades

[role="_abstract"]
You can accelerate upgrades in large clusters by partitioning worker nodes into custom Machine Config Pools (MCPs) that align with Kubernetes failure domains (KFDs). This approach changes the default sequential node-by-node update process into a parallel partition-based strategy.

This strategy is particularly relevant for bare metal and on-premise environments where you cannot add temporary surge capacity during upgrades. By updating all nodes in a single failure domain simultaneously, you can reduce upgrade time while keeping other failure domains available to maintain High Availability (HA) for workloads.

Configure each custom MCP with `maxUnavailable: 100%` so that all nodes in that pool to update at the same time. This setting applies only to the nodes in the selected MCP, not to the entire cluster.

Plan the number and size of custom MCPs based on your cluster topology, workload distribution, and application availability requirements, including Pod Disruption Budgets (PDBs).

> 

Reviewer 2: I believe we need some additional context here. Things not covered are:
* Use of zones is not a guarantee that the scheduler will spread replicas across zones. It is a soft constraint, but it is beneficial to have the scheduler taking failure domains (and thus upgrade domains) into account
* The ability to take an entire failure domain offline depends on several criteria
  * The application/workload is designed to be highly available
  * The application/workload has PodDisruptionBudgets protecting the HA replicas
  * The application is able to meet minimum service level requirements while the failure domain is offline
* Recommendation is to have sufficient spare capacity in the cluster to accomodate pods disrupted when failure domain is taken offline
* If application level contraints require less than a full failure domain be taken offline concurrently the maxUnavailable setting can be reduced, eg to 50% of the failure domain, or whatever capacity allows the application to meet service requirements.

Author: Added the suggested content.

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have the OpenShift CLI (`oc`) installed.
* You have identified the failure domains (zones) in your cluster topology (for example, `worker-0`, `worker-1`, `worker-2`, `worker-3`).
* You have ensured that the cluster has sufficient spare capacity to support the workload if one failure domain is unavailable.

.Procedure

. Create a YAML file for each custom MCP that corresponds to a failure domain in your cluster, as in the following example:
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-0
  labels:
    machineconfiguration.openshift.io/role: worker-0
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values: [ worker, worker-0 ]

>   paused: true

Reviewer 2: The normal state of pause is `false`. Only set it to true when the upgrade process is starting. Set it true **before** starting the control plane upgrade.

Author: Updated as follows:
"paused: Set this value to `false` during normal cluster operation. Before starting the upgrade process, set this value to `true` for each custom MCP so that worker pool updates do not begin automatically during the control plane upgrade. Unpause each custom MCP only when you are ready to upgrade that specific failure domain."

+
[source,terminal]
----
$ oc get mcp -o custom-columns=NAME:.metadata.name,MACHINECOUNT:.status.machineCount
----
+
.Example output
[source,terminal]
----
NAME       MACHINECOUNT
master     3
worker     0
worker-0   2
worker-1   2
worker-2   2
worker-3   2
----

. Verify that the worker nodes are correctly assigned to the custom MCPs and that the pools are created successfully.
+
[source,terminal]
----
$ oc get nodes --show-labels=false -L 'topology.kubernetes.io/zone'
----
+
.Example output
----
NAME                                        STATUS   ROLES                  AGE   VERSION   ZONE
master-0.topology.kubernetes.io/zone=kfd0   Ready    control-plane,master   27h   v1.31.13
worker-0.topology.kubernetes.io/zone=kfd0   Ready    worker,worker-0        27h   v1.31.13  kfd0
worker-1.topology.kubernetes.io/zone=kfd1   Ready    worker,worker-1        27h   v1.31.13  kfd0
worker-2.topology.kubernetes.io/zone=kfd2   Ready    worker,worker-2        27h   v1.31.13  kfd2

> worker-3.topology.kubernetes.io/zone=kfd3   Ready    worker,worker-3        27h   v1.31.13  kfd3

Reviewer 2: As currently shows this has one worker in each MCP and thus one worker in each zone. I think it would be more instructive to show 2 workers per MCP and zone, even if we use only 2 MCPs if we need to keep this short.

Author: Can we update it as follows:

.Example output
----
NAME       STATUS   ROLES             AGE   VERSION    ZONE
worker-0   Ready    worker,worker-0   27h   v1.31.13  kfd0
worker-1   Ready    worker,worker-0   27h   v1.31.13  kfd0
worker-2   Ready    worker,worker-1   27h   v1.31.13  kfd1
worker-3   Ready    worker,worker-1   27h   v1.31.13  kfd1
----

Reviewer: I think @imiller0 is referring to:

- 2 nodes per KFD, e.g. worker-00 and worker-01 for kfd0 and so on.
- Each KFD matches a MCP, so there are worker-0 to worker-3 MCPs, then it should be kfd0 to kfd3.

TLDR;

```
NAME       STATUS   ROLES             AGE   VERSION    ZONE
worker-00   Ready    worker,worker-0   27h   v1.31.13  kfd0
worker-01   Ready    worker,worker-0   27h   v1.31.13  kfd0
worker-10   Ready    worker,worker-1   27h   v1.31.13  kfd1
worker-11   Ready    worker,worker-1   27h   v1.31.13  kfd1
worker-20   Ready    worker,worker-1   27h   v1.31.13  kfd2
worker-21   Ready    worker,worker-1   27h   v1.31.13  kfd2
worker-30   Ready    worker,worker-1   27h   v1.31.13  kfd3
worker-31   Ready    worker,worker-1   27h   v1.31.13  kfd3
```

Author: Updated as per the suggestion.

----

> . Schedule workloads on the cluster only after you verify that nodes are labeled correctly and distributed across Kubernetes failure domains.

Reviewer 2: Should advise that the consequence of scheduling before zone labeling is that the scheduler won't distribute across zones

> . Keep each custom MCP paused until you are ready to upgrade the cluster.

Reviewer 2: As noted above, the normal state for pause should be `false`. The wording here should be pause the MCPs when you are ready to upgrade the cluster

Author: Updated as follows:
". Pause each custom MCP before starting the cluster upgrade so that worker pool updates do not begin automatically during the control plane upgrade."

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have the OpenShift CLI (`oc`) installed.
* You have identified the failure domains (zones) in your cluster topology (for example, `worker-0`, `worker-1`, `worker-2`, `worker-3`).
* You have ensured that the cluster has sufficient spare capacity to support the workload if one failure domain is unavailable.

.Procedure

. Create a YAML file for each custom MCP that corresponds to a failure domain in your cluster, as in the following example:
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: worker-0
  labels:
    machineconfiguration.openshift.io/role: worker-0
spec:
  machineConfigSelector:
    matchExpressions:
      - key: machineconfiguration.openshift.io/role
        operator: In
        values: [ worker, worker-0 ]
  paused: false
  maxUnavailable: 100%
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/worker-0: ""
----
+
Ensure the following configurations are present:

> * *maxUnavailable*: Set this value to `100%`. This setting enables the Machine Config Operator (MCO) to update all nodes in the single failure domain simultaneously, significantly reducing upgrade time for the zone.

Reviewer 3: ```suggestion
* *maxUnavailable*: Set this value to `100%`. This setting defines either an integer number or percentage of nodes in the pool that can go Unavailable during an update, significantly reducing upgrade time for the MCP.
```
https://docs.redhat.com/en/documentation/openshift_container_platform/4.22/html/machine_apis/machineconfigpool-machineconfiguration-openshift-io-v1

> * *paused*: Set this value to `false` during normal cluster operation. Before starting the upgrade process, set this value to `true` for each custom MCP so that worker pool updates do not begin automatically during the control plane upgrade. Unpause each custom MCP only when you are ready to upgrade that specific failure domain.

Reviewer 3: ```suggestion
* *paused*: Set this value to `false` during normal cluster operation. Before starting the upgrade process, set this value to `true` for each custom MCP so that worker pool updates do not begin automatically during the control plane upgrade. Unpause each custom MCP only when you are ready to upgrade that specific MCP.
```

worker-01   Ready    worker,worker-0   27h   v1.31.13  kfd0
worker-10   Ready    worker,worker-1   27h   v1.31.13  kfd1
worker-11   Ready    worker,worker-1   27h   v1.31.13  kfd1
worker-20   Ready    worker,worker-1   27h   v1.31.13  kfd2
worker-21   Ready    worker,worker-1   27h   v1.31.13  kfd2
worker-30   Ready    worker,worker-1   27h   v1.31.13  kfd3
worker-31   Ready    worker,worker-1   27h   v1.31.13  kfd3
----

. Schedule workloads on the cluster only after you verify that nodes are labeled correctly and distributed across Kubernetes failure domains.

. Schedule workloads on the cluster only after you verify that nodes are labeled correctly and distributed across Kubernetes failure domains. If workloads are scheduled before nodes are labeled with failure domains, the Kubernetes scheduler cannot consider zone distribution when placing application replicas.

. Pause each custom MCP before starting the cluster upgrade so that worker pool updates do not begin automatically during the control plane upgrade.
+
[source,terminal]
----
$ oc get mcp
----

. Upgrade the control plane by using the standard cluster upgrade process.
+
[source,terminal]
----
$ oc adm upgrade
----

> . Unpause one custom MCP at a time to begin upgrading that failure domain.

Reviewer 3: ```suggestion
. Unpause one custom MCP at a time to begin upgrading.
```
